### PR TITLE
MaxwellDMA: Implement semaphore operations

### PR DESCRIPTION
**Pull request by asLody**
This implements the missing semaphore for MaxwellDMA.
- Used by BOTW & PLA.

### CODE REVIEW

**Review by bunnei — APPROVED** [review]

**Comment by bunnei** [comment]
LGTM, thank you!

File: src/video_core/engines/maxwell_dma.cpp [hunk 1]
```diff
@@ -244,4 +244,23 @@ void MaxwellDMA::FastCopyBlockLinearToPitch() {
     memory_manager.WriteBlock(regs.offset_out, write_buffer.data(), dst_size);
 }
 
+void MaxwellDMA::ReleaseSemaphore() {
+    const auto type = regs.launch_dma.semaphore_type;
+    if (type == LaunchDMA::SemaphoreType::NONE) {
+        return;
+    }
```
Comment (v1993) [inline src/video_core/engines/maxwell_dma.cpp:251]: Nitpick - this probably can be merged into switch-case below for code readability without real performance impact.

Comment (v1993) [inline src/video_core/engines/maxwell_dma.cpp:251]: Nitpick - this probably can be merged into switch-case below for code readability without real performance impact.

**Comment by goldenx86** [comment]
Hi, can I mention you in the next yuzu Progress Report for your work on this PR?

**Review by FernandoS27 — APPROVED** [review]

**Comment by asLody** [comment]
> Hi, can I mention you in the next yuzu Progress Report for your work on this PR?

Sure!

**Comment by edisionnano** [comment]
@asLody Very interesting PR but can you explain a bit more thoroughly what it does? Also please demonstrate what exactly it fixes on the two games you mentioned, BOTW and PLA.

**Comment by FernandoS27** [comment]
> @asLody Very interesting PR but can you explain a bit more thoroughly what it does? Also please demonstrate what exactly it fixes on the two games you mentioned, BOTW and PLA.

This is a very simple semaphore implementation. They are all over the code just that we likely never triggered this anywhere and he saw the asserts.
 What it does have to do with syncing and the game knowing that the dma operation has been finished (normally they use a query in maxwell3d(nouveau name) threed(nvidia's name) engine but some subroutines on pbdma may use this semaphore as well.
 
 The offcial documentation about semaphores can be found here: https://github.com/NVIDIA/open-gpu-doc/blob/master/manuals/volta/gv100/dev_pbdma.ref.txt#L1520
 
As well as the official DMA engine documentation is here https://github.com/NVIDIA/open-gpu-doc/blob/master/classes/dma-copy/clb0b5.h

**Comment by edisionnano** [comment]
> > @asLody Very interesting PR but can you explain a bit more thoroughly what it does? Also please demonstrate what exactly it fixes on the two games you mentioned, BOTW and PLA.
> 
> This is a very simple semaphore implementation. They are all over the code just that we likely never triggered this one ins specific anywhere and he saw the asserts. What it does have to do with syncing and the game knowing that the dma operation has been finished (normally they use a query in maxwell3d(nouveau name) threed(nvidia's name) engine but some subroutines on pbdma may use this semaphore as well.
> 
> The offcial documentation about semaphores can be found here: https://github.com/NVIDIA/open-gpu-doc/blob/master/manuals/volta/gv100/dev_pbdma.ref.txt#L1520
> 
> the semaphores documentation is Volta's but hardware tests have shown they have not changed much, except for a few configurations with 64 bits reductions that volta+ support
> 
> As well as the official DMA engine documentation (for maxwell) is here https://github.com/NVIDIA/open-gpu-doc/blob/master/classes/dma-copy/clb0b5.h

I see, thanks Blink! What exactly does it improve on BOTW and PLA?

**Comment by FernandoS27** [comment]
> > > @asLody Very interesting PR but can you explain a bit more thoroughly what it does? Also please demonstrate what exactly it fixes on the two games you mentioned, BOTW and PLA.
> > 
> > 
> > This is a very simple semaphore implementation. They are all over the code just that we likely never triggered this one ins specific anywhere and he saw the asserts. What it does have to do with syncing and the game knowing that the dma operation has been finished (normally they use a query in maxwell3d(nouveau name) threed(nvidia's name) engine but some subroutines on pbdma may use this semaphore as well.
> > The offcial documentation about semaphores can be found here: https://github.com/NVIDIA/open-gpu-doc/blob/master/manuals/volta/gv100/dev_pbdma.ref.txt#L1520
> > the semaphores documentation is Volta's but hardware tests have shown they have not changed much, except for a few configurations with 64 bits reductions that volta+ support
> > As well as the official DMA engine documentation (for maxwell) is here https://github.com/NVIDIA/open-gpu-doc/blob/master/classes/dma-copy/clb0b5.h
> 
> I see, thanks Blink! What exactly does it improve on BOTW and PLA?

To be honest, I'm not certain. There was some BS drama around this PR that I had to deal with, so I spoke with Skyline emulator devs about what it is about and stuffs. In BotW, normally the game measures frame time using "report queries" this are "semaphores" whose only purpose is to write the timestamp. Deko3D has a version of this. From what I understood that game uses this particular semaphore as well for getting 1 of those timestamps.

About PLA, no idea, likely he implemented this just based on the ASSERT showing up in yuzu's log as the implementation is actually not correct if you read the link I passed you. The timestamp is supposed to be written first. I merged regardless as the other semaphores are broken like that. I intend to fix this in YFC.

**Comment by edisionnano** [comment]
> > > > @asLody Very interesting PR but can you explain a bit more thoroughly what it does? Also please demonstrate what exactly it fixes on the two games you mentioned, BOTW and PLA.
> > > 
> > > 
> > > This is a very simple semaphore implementation. They are all over the code just that we likely never triggered this one ins specific anywhere and he saw the asserts. What it does have to do with syncing and the game knowing that the dma operation has been finished (normally they use a query in maxwell3d(nouveau name) threed(nvidia's name) engine but some subroutines on pbdma may use this semaphore as well.
> > > The offcial documentation about semaphores can be found here: https://github.com/NVIDIA/open-gpu-doc/blob/master/manuals/volta/gv100/dev_pbdma.ref.txt#L1520
> > > the semaphores documentation is Volta's but hardware tests have shown they have not changed much, except for a few configurations with 64 bits reductions that volta+ support
> > > As well as the official DMA engine documentation (for maxwell) is here https://github.com/NVIDIA/open-gpu-doc/blob/master/classes/dma-copy/clb0b5.h
> > 
> > 
> > I see, thanks Blink! What exactly does it improve on BOTW and PLA?
> 
> To be honest, I'm not certain. There was some BS drama around this PR that I had to deal with, so I spoke with Skyline emulator devs about what it is about and stuffs. In BotW, normally the game measures frame time using "report queries" this are "semaphores" whose only purpose is to write the timestamp. Deko3D has a version of this. From what I understood that game uses this particular semaphore as well for getting 1 of those timestamps. This issue, we have had been talking for a while but I had not really looked into it.
> 
> About PLA, no idea, likely he implemented this just based on the ASSERT showing up in yuzu's log as the implementation is actually not correct if you read the link I passed you. The timestamp is supposed to be written first. I merged regardless as the other semaphores are broken like that. I intend to fix this in YFC.

Thanks for the explanation blink. I assume deko alsl uses the nouveu impl?

**Pull request merged**